Improve use of tempfile in tests

aaugustin opened the pull request:
I have larger changes coming in this area, but I'd like to start by merging these two commits.

They're self-contained and valuable in themselves.

timgraham commented:
There is a problem on Windows with your branch:
```
======================================================================
ERROR: test_deleted_removed (utils_tests.test_autoreload.TestFilenameGenerator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\Users\Tim\code\django\tests\utils_tests\test_autoreload.py", line 97,
 in test_deleted_removed
    import_module(modulename)
  File "c:\Users\Tim\.envs\django342\lib\importlib\__init__.py", line 109, in im
port_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 2254, in _gcd_import
  File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
  File "<frozen importlib._bootstrap>", line 2226, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1467, in exec_module
  File "<frozen importlib._bootstrap>", line 1571, in get_code
  File "<frozen importlib._bootstrap>", line 1623, in get_data
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\Tim\\AppData\\Local\\
Temp\\tmp7zfh1vtt.py'

----------------------------------------------------------------------
```

timgraham commented:
It seems this change results in leftover temporary directories that 5de31cb8 was added to prevent. I see four `tmpXXXX` directories and `test_project` in my `/tmp` after the test suite completes with your branch.

aaugustin commented:
The commit I just pushed fixes the problem with leftover temporary directories. I'll squash it into the first one before merging.

aaugustin commented:
And I pushed another commit which should take care of the test failure on Windows. It's the usual "NamedTemporaryFile returns a file name but you can't reopen the file". I could have used the version in `django.core.files.temp` but it's just as simple to skip the problem entirely.

timgraham commented on tests/utils_tests/test_autoreload.py:
```diff
@@ -88,12 +89,12 @@ def test_only_new_files(self):
-            os.close(fd)
-            os.remove(filepath)
-        self.assertNotIn(filepath, gen_filenames())
+        dirname = tempfile.mkdtemp()
```
Is your thinking that even if the assertion fails, we don't care about cleaning up after ourself since we're in the temp directory that will be deleted at the end of the test run anyway?

aaugustin replied:
Indeed, I didn't bother removing the temporary directory (always) and the temporary file (only if the first assertion fails) because that will happen automatically at the end of the test run.

timgraham commented:
Looks good and works on Linux & Windows.

timgraham commented:
Actually, there are some problems on Python 2 on Windows:
```
======================================================================
ERROR: test_staticfiles_dirs_subdir (staticfiles_tests.tests.TestFindStatic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python27\Lib\shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "C:\Python27\Lib\shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "C:\Python27\Lib\shutil.py", line 250, in rmtree
    os.remove(fullname)
WindowsError: [Error 123] The filename, directory name, or volume label syntax i
s incorrect: 'c:\\users\\tim\\appdata\\local\\temp\\django_ub0amb\\tmpjgqlvp\\te
st\\?.txt'
```

aaugustin commented:
I believe I introduced this failure when I removed `ignore_errors=True` at the same time as I dropped `rmtree_errorhandler`. Using the same workaround as in `runtests.teardown` should probably fix it:

    # Removing the temporary TEMP_DIR. Ensure we pass in unicode
    # so that it will successfully remove temp trees containing
    # non-ASCII filenames on Windows. (We're assuming the temp dir
    # name itself does not contain non-ASCII characters.)
    shutil.rmtree(six.text_type(TEMP_DIR))

However, if we embrace the idea that runtests takes care of creating and disposing of the temporary directory, we can simply remove the cleanup code.

What do you think?

aaugustin commented:
For now just fixing the test failure will allow me to move on. My test parallelization project is already taking me way too far...

timgraham commented:
It works. Being lax with cleanup and relying on runtests seems okay.

aaugustin commented:
Phew. Thanks a lot for testing.
